README.md
# SPX

Unofficial Edimax Smartplug Libary. Control SP-1101W and SP-2101W from Python.

## Install
spx is available on [pypi](https://pypi.python.org/pypi/spx). Install it using pip:
```
pip install spx
```



## Usage as a Library

```
import spx

p = spx.SmartPlug('ip_or_hostname')
p.set_state(bool)
p.switch(bool)  # alias for set_state(bool)
p.on()  # alias for set_state(True)
p.off() # alias for set_state(False)
p.get_usage()  # SP-2101W only
p.monitor() #SP-2101W only
#as a convenience all commands are available as functions as well:
spx.on('ip_or_hostname') etc.
```

## Usage of spx.tool
```
$ python -m spx.tool --help
# or
$ spx-tool --help
```

CHANGELOG.md
# 0.20 (Broken Arrow - lot's of untested changes)
* get_usage returns correct types, not strings
* Tests broken
* Add first Control Strategy (untested) 
* Move monitor from Library to SPX Tool


tests.py
import datetime
import unittest
from unittest.mock import MagicMock, patch
from spx import Smartplug


class SmartplugTestCase(unittest.TestCase):
    @patch('spx.Smartplug._get_timestamp', return_value=datetime.datetime(2015,9,17,14,28,1))
    @patch('requests.post', return_value=MagicMock())
    def test_get_usage(self, p, d):
        p.return_value.status_code = 200
        p.return_value.content = '<?xml version="1.0" encoding="UTF8"?><SMARTPLUG id="edimax">\n<CMD id="get">\n    <NOW_POWER><Device.System.Power.LastToggleTime>20150912160852</Device.System.Power.LastToggleTime><Device.System.Power.NowCurrent>0.0450</Device.System.Power.NowCurrent><Device.System.Power.NowPower>5.29</Device.System.Power.NowPower><Device.System.Power.NowEnergy.Day>0.080</Device.System.Power.NowEnergy.Day><Device.System.Power.NowEnergy.Week>0.080</Device.System.Power.NowEnergy.Week><Device.System.Power.NowEnergy.Month>1.214</Device.System.Power.NowEnergy.Month>\n    </NOW_POWER>\n</CMD>\n</SMARTPLUG>'
        plug = Smartplug('240.0.0.1')

        self.assertDictEqual(
            plug.get_usage(),
            {
                't': datetime.datetime(2015, 9, 17, 14, 28, 1),
                'current': 0.045,
                'day': 0.080,
                'last_toggle_time': datetime.datetime(2015, 9, 12, 16, 8, 52),
                'month': 1.214,
                'power': 5.29,
                'week': 0.080,
            }
        )
        p.assert_called_once()
        self.assertEquals(p.call_args[0][0], 'http://admin:1234@240.0.0.1:10000/smartplug.cgi')
        self.assertEquals(p.call_args[1]['data'], '<?xml version="1.0" encoding="UTF8"?><SMARTPLUG id="edimax"><CMD id="get"><NOW_POWER></NOW_POWER></CMD></SMARTPLUG>')

    @patch('requests.post', return_value=MagicMock())
    def test_set_state(self, p):
        p.return_value.status_code = 200
        p.return_value.content = '<?xml version="1.0" encoding="UTF8"?><SMARTPLUG id="edimax"></SMARTPLUG>'
        self.fail('Invalid Response')
        plug = Smartplug('240.0.0.1')

        self.assertTrue(plug.set_state('on'))

    @patch('requests.post', return_value=MagicMock())
    def test_get_state(self, p):
        p.return_value.status_code = 200
        p.return_value.content = '<?xml version="1.0" encoding="UTF8"?><SMARTPLUG id="edimax"></SMARTPLUG>'
        self.fail('Invalid Response')


edimax.py
#!/usr/bin/env python
import datetime
import requests
import re
import sys
import threading

HEIZUNG_URL='http://admin:1234@192.168.178.5:10000/smartplug.cgi'
ON='<?xml version="1.0" encoding="UTF8"?><SMARTPLUG id="edimax"><CMD id="setup"><Device.System.Power.State>ON</Device.System.Power.State></CMD></SMARTPLUG>'
OFF='<?xml version="1.0" encoding="UTF8"?><SMARTPLUG id="edimax"><CMD id="setup"><Device.System.Power.State>OFF</Device.System.Power.State></CMD></SMARTPLUG>'
QUERY='<?xml version="1.0" encoding="UTF8"?><SMARTPLUG id="edimax"><CMD id="get"><NOW_POWER><Device.System.Power.NowPower></Device.System.Power.NowPower></NOW_POWER></CMD></SMARTPLUG>'

def get_wattage(edimax_url):
    response = requests.post(edimax_url,data=QUERY)
    if response.status_code != 200:
        print('error', response)
        sys.exit(1)
    watt = re.search(br'\<Device\.System\.Power\.NowPower\>(?P<watt>.*)\<\/Device\.System\.Power\.NowPower\>',response.content).groupdict()['watt']
    return watt.decode('utf-8')

def print_watt(repeat=0):
    if(repeat!=0):
        threading.Timer(repeat, lambda: print_watt(repeat)).start()
    print('{timestamp} {wattage}'.format(timestamp=datetime.datetime.now().isoformat(), wattage=get_wattage(HEIZUNG_URL)))
    sys.stdout.flush()

def main():
    print_watt(int(sys.argv[1]))

if __name__ == '__main__':
    main()


spx/tool.py
#!/usr/bin/env python
import datetime
import sys
import argparse
import logging
from apscheduler.schedulers.blocking import BlockingScheduler

import spx


class SPXTool(object):
    DATETIME_FORMAT = '%Y%m%d%H%M%S'
    DEFAULT_INTERVAL = 5

    FORMATS = {
        'minimal': '{t:' + DATETIME_FORMAT + '} {power}',
        'default': 't:' + DATETIME_FORMAT +'} {current} {power} {last_toggle_time:' + DATETIME_FORMAT + '}',
    }

    @staticmethod
    def run():
        SPXTool()()

    def __init__(self):
        self.args = None
        self.parse_args()
        self.smartplug = spx.Smartplug(self.args.host,
                                       self.args.username, self.args.password)

    def __call__(self):
        self.args.func()

    def run_monitor(self):
        usage = self.smartplug.get_usage()
        print('{t:' + self.DATETIME_FORMAT + '} {power}'.format(
            **usage
        ), flush=True)

    def monitor(self, interval=None):
        if interval is None:
            interval = self.DEFAULT_INTERVAL
        if type(interval) is str:
            interval = int(interval)
        scheduler = BlockingScheduler()
        self.run_monitor()
        if interval < 1:
            return

        scheduler.add_job(self.run_monitor, 'interval', seconds=int(interval))

        try:
            scheduler.start()
        except KeyboardInterrupt:
            sys.exit(0)

    def parse_args(self):
        common_parser = argparse.ArgumentParser(add_help=False)

        c_group = common_parser.add_argument_group('Credentials')
        c_group.add_argument(
            '-u', '--username',
            default=spx.Smartplug.DEFAULT_USERNAME,
            help='Username to authenticate against Smartplug'
        )

        c_group.add_argument(
            '-p', '--password',
            default=spx.Smartplug.DEFAULT_PASSWORD,
            help='Password to authenticate against Smartplug'
        )

        l_group = common_parser.add_argument_group('Logging')
        l_group.add_argument(
            '-l', '--loglevel',
            choices=['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'],
            default='ERROR',
            help='Log level'
        )

        parser = argparse.ArgumentParser(parents=[common_parser])
        parser.add_argument('host', help='Hostname or IP of the Smartplug')

        commands = parser.add_subparsers(
            title='Commands',
            help='Description'
        )
        get_state_command = commands.add_parser(
            'get_state', parents=[common_parser],
            help='Get state of Smartplug (on or off)'
        )
        get_state_command.set_defaults(func=lambda: self.smartplug.get_state())

        set_state_command = commands.add_parser(
            'set_state', parents=[common_parser],
            help='Set state of Smartplug (on or off)'
        )
        set_state_command.add_argument(
            'state',
            type=str.lower,
            choices=spx.Smartplug.BOOLEAN_STATES.keys(),
        )
        set_state_command.set_defaults(func=lambda: self.smartplug.set_state(self.args.state))

        on_command = commands.add_parser(
            'on', parents=[common_parser],
            help='Alias for set_state on'
        )
        on_command.set_defaults(func=lambda: self.smartplug.on())

        off_command = commands.add_parser(
            'off', parents=[common_parser],
            help='Alias for set_state off'
        )
        off_command.set_defaults(func=lambda: self.smartplug.off())

        switch_command = commands.add_parser(
            'switch', parents=[common_parser], help='Alias for set_state'
        )
        switch_command.add_argument(
            'state', type=str.lower, choices=spx.Smartplug.BOOLEAN_STATES.keys(),)
        switch_command.set_defaults(func=lambda: self.smartplug.set_state(self.args.state))

        monitor_command = commands.add_parser('monitor', parents=[common_parser],
            help='Monitor power usage'
        )
        monitor_command.add_argument(
            '-i', '--interval', type=int, default=self.DEFAULT_INTERVAL,
            help='monitor interval in seconds'
        )
        monitor_command.add_argument(
            '-f', '--format', default=self.FORMATS['default'],
        )

        monitor_command.set_defaults(
            func=lambda: self.monitor(self.args.interval)
        )
        self.args = parser.parse_args()

        logging.basicConfig(level=getattr(logging, self.args.loglevel),
                            stream=sys.stderr)


if __name__ == '__main__':
    SPXTool()()


spx/strategy/simple_strategy.py
import collections
import logging
from spx import Smartplug
from .strategy_interface import StrategyInterface

log = logging.getLogger(__name__)


class SimpleStrategy(StrategyInterface):
    """
    A very simple control strategy. It could be named the
    DidYouTryTurningItOffAndOnAgainStrategy.

    If a consumer never draws more than a given amount of power during a
    given time frame it is considered to be in errornous state and therefore
    power cycled.

    This strategy won't act until update has been called history_size times

    :param history_size the data points added through this many calls to update
    are considered in the Strategy
    :param limit the minimum amount power in watts that has to be drawn for the
    device to be considered active.
    :param cycle_duration stay off for this many seconds on error.

    """
    def __init__(self, history_size, limit, cycle_duration):
        self.history = collections.deque(maxlen=history_size)
        self.limit = limit
        self.cycle_duration = cycle_duration
        self.in_cycle = 0

    def update(self, d):
        self.history.append(d)

    def run(self, plug: Smartplug):
        if len(self.history) != self.history.maxlen:
            return

        if (not self.in_cycle
                and max(x.get('power', 0.0) for x in self.history) < self.limit):
            plug.off()
            self.in_cycle = self.cycle_duration
        else:
            self.in_cycle -= 1
            if self.in_cycle == 0:
                plug.on()


spx/smartplug.py
# -*- coding: utf-8 -*-
import datetime
import logging
import re
import textwrap
from xml.dom.minidom import parseString

import requests

log = logging.getLogger(__name__)


def un_camel(s):
    return re.sub(r'((?<=[a-z0-9])[A-Z]|(?!^)[A-Z](?=[a-z]))',
                  r'_\1', s).lower()


class SmartplugException(Exception):
    pass


class SmartplugCommandFailed(SmartplugException):
    pass


class Smartplug(object):
    DEFAULT_USERNAME = 'admin'
    DEFAULT_PASSWORD = '1234'
    DATETIME_FORMAT = '%Y%m%d%H%M%S'


    URL = 'http://{p.username}:{p.password}@{p.host}:10000/smartplug.cgi'

    MESSAGE = textwrap.dedent('''
        <?xml version="1.0" encoding="UTF8"?>
        <SMARTPLUG id="edimax">{command}
        </SMARTPLUG>''')

    SET_STATE_MESSAGE = MESSAGE.format(command=textwrap.dedent('''
        <CMD id="setup">
            <Device.System.Power.State>{state}</Device.System.Power.State>
        </CMD>'''))

    GET_STATE_MESSAGE = MESSAGE.format(command=textwrap.dedent('''
        <CMD id="get">
            <Device.System.Power.State></Device.System.Power.State>
        </CMD>'''))

    GET_USAGE_MESSAGE = MESSAGE.format(command=textwrap.dedent('''
        <CMD id="get">
            <NOW_POWER>
            </NOW_POWER>
        </CMD>'''))

    BOOLEAN_STATES = {'1': True, 'yes': True, 'true': True, 'on': True,
                      '0': False, 'no': False, 'false': False, 'off': False}

    def __init__(self, host, username=None, password=None):
        self.host = host
        self.username = username if username is not None else self.DEFAULT_USERNAME
        self.password = password if password is not None else self.DEFAULT_PASSWORD

        self.url = self.URL.format(p=self)

    def _convert_to_boolean(self, value):
        """
        Return a boolean value translating from other types if necessary.
        (Taken from configparser)
        """
        if type(value) is bool:
            return value
        if value.lower() not in self.BOOLEAN_STATES:
            raise ValueError('Not a boolean: %s' % value)
        return self.BOOLEAN_STATES[value.lower()]

    def _send_command(self, command):
        log.debug(command)
        response = requests.post(self.url, data=command)

        if response.status_code != 200:
            raise SmartplugCommandFailed(response.status_code,
                                         response.content)

        log.debug(response.content)
        return response.content

    def get_state(self):
        self._send_command(self.GET_STATE_MESSAGE)
        raise NotImplemented()

    def set_state(self, state):
        self._send_command(self.SET_STATE_MESSAGE.format(
            state='ON' if self._convert_to_boolean(state) else 'OFF'
        ))
        return True

    def switch(self, state):
        return self.set_state(state)

    def on(self):
        return self.set_state(True)

    def off(self):
        return self.set_state(False)

    def _get_timestamp(self):
        return datetime.datetime.now()

    def get_usage(self):
        response = self._send_command(self.GET_USAGE_MESSAGE)
        dom = parseString(response)
        dom_data = dom.getElementsByTagName('NOW_POWER')
        result = {
            't': self._get_timestamp()
        }
        for node in dom_data[0].childNodes:
            if node.nodeType != node.ELEMENT_NODE:
                continue
            key = un_camel(re.sub('.*\.(Now)?', '', node.nodeName))
            raw_value = node.firstChild.nodeValue.strip()
            if key == 'last_toggle_time':
                value = datetime.datetime.strptime(raw_value, self.DATETIME_FORMAT)
            else:
                value = float(raw_value)
            result[key] = value
        return result


def on(host, username=None, password=None):
    return Smartplug(host, username=username, password=password).on()


def off(host, username=None, password=None):
    return Smartplug(host, username=username, password=password).off()


def get_state(host, username=None, password=None):
    return Smartplug(host, username=username, password=password).get_state()


def set_state(host, state, username=None, password=None):
    return Smartplug(host, username=username, password=password).set_state(
        state)


def switch(host, state, username=None, password=None):
    return Smartplug(host, username=username, password=password).switch(state)


def get_usage(host, username=None, password=None):
    return Smartplug(host, username=username, password=password).get_usage()
